drivers/ksz8851snl/ksz8851_netdev.c

```diff
@@ -0,0 +1,622 @@
+		else
+			cmd |= IO_REG_READ | KSZ8851_CMD_B0;
+	}
+	char cmdChain[3] = { (cmd >> 8), (cmd & 0xff), 0 };
+
+	char result[3];
+
+	ksz_spi_transfer(dev, cmdChain, result, 3);
+
+	return result[2];
+}
+
+
+static void ksz8851snl_get_mac_addr(netdev2_t *encdev, uint8_t* buf)
+{
+	ksz8851snl_t * dev = (ksz8851snl_t*)encdev;
+
+	lock(dev);
+	for (uint8_t x = 0; x < 6; x++)
+		*buf++ = reg_get8(dev, 0x15 - x);
+	unlock(dev);
+}
+
+static int ksz8851snl_get_iid(netdev2_t *netdev, eui64_t *value, size_t max_len)
+{
+	if (max_len < sizeof(eui64_t))
+		return -EOVERFLOW;
+
+	uint8_t addr[ETHERNET_ADDR_LEN];
+	ksz8851snl_get_mac_addr(netdev, addr);
+	ethernet_get_iid(value, addr);
+
+	return sizeof(eui64_t);
+}
+
+static int ksz8851snl_get(netdev2_t *dev, netopt_t opt, void *value, size_t max_len)
+{
+	int res = -EINVAL;
+	uint16_t *tgt = (uint16_t *)value;
+
+	switch (opt) {
+		case NETOPT_CHANNEL: //0
+		case NETOPT_IS_CHANNEL_CLR: //1
+			res = -EINVAL;
+			break;
+		case NETOPT_ADDRESS: //2
+			if (max_len < ETHERNET_ADDR_LEN)
+				res = -EINVAL;
+			else {
+				ksz8851snl_get_mac_addr(dev, (uint8_t*)value);
+				res = ETHERNET_ADDR_LEN;
+			}
+			break;
+		case NETOPT_ADDRESS_LONG: //3
+			res = -EINVAL;
+			break;
+		case NETOPT_ADDR_LEN: //4
+		case NETOPT_SRC_LEN: //5
+			assert(max_len == 2);
+			*tgt = 6;
+			res = sizeof(uint16_t);
+			break;
+		case NETOPT_NID: //6
+			res = -EINVAL;
+			break;
+		case NETOPT_IPV6_IID: //7
+			return ksz8851snl_get_iid(dev, value, max_len);
+		case NETOPT_TX_POWER: //8
+		case NETOPT_MAX_PACKET_SIZE: //9
+		case NETOPT_PRELOADING: //10
+		case NETOPT_PROMISCUOUSMODE:  //11
+		case NETOPT_AUTOACK: //12
+		case NETOPT_RETRANS: //13
+			break;
+		case NETOPT_PROTO:  //14
+			res = -EINVAL;
+			break;
+		case NETOPT_STATE: //15
+			res = 0;
+			*tgt = NETOPT_STATE_IDLE;
+			break;
+		case NETOPT_RAWMODE:  //16
+		case NETOPT_RX_START_IRQ:  //17
+		case NETOPT_RX_END_IRQ:  //18
+		case NETOPT_TX_START_IRQ:  //19
+		case NETOPT_TX_END_IRQ:  //20
+		case NETOPT_AUTOCCA:  //21
+		case NETOPT_CSMA:  //22
+		case NETOPT_CSMA_RETRIES: //23
+			res = -EINVAL;
+			break;
+		case NETOPT_IS_WIRED: //24
+			res = 1;
+			break;
+		case NETOPT_DEVICE_TYPE: //25
+			*tgt = NETDEV2_TYPE_ETHERNET;
+			res = 2;
+		 	break;
+        case NETOPT_CHANNEL_PAGE: //26
+            res = -EINVAL;
+            break;
+		default:
+			DEBUG(SECT_NAME"Device Get Opt Called and not handled: %d\n", opt);
+			res = -ENOTSUP;
+			break;
+	}
+
+	return res;
+}
+
+static int ksz8851snl_set(netdev2_t *dev, netopt_t opt, void *value, size_t value_len)
+{
+	int res = 0;
+
+	switch (opt) {
+	default:
+		printf("Device Set Opt Called: %d\n", opt);
+		return -ENOTSUP;
+	}
+
+	return res;
+}
+
+/*
+ * Process for transmitting packet
+ * 1. Check if the send buffer can hold the packet register: TXMIR
+ * 1b. TODO: If there is not room on the KSZ buffer you can setup an interrupt
+ *           for when there is room left
```

Comment:
thinking out loud: wouldn't it make sense for know, to just actively scan for the TX busy bit (if there is one) before sending to make sure the previous transmission is finished? In the worst case, on thread would do this, all other waiting threads would be put away by the mutex. And the chance that there is still an ongoing transmission is not very high - most likely the ethernet will send out frames much faster than our relatively slow microcontrollers can provide them with new ones, right?